File: trwm/projection.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Iterable, Mapping

from .core import stable_hash


PROJECTION_CONTRACT_SCHEMA = "trwm.projection_contract.v1"
PROJECTION_MANIFEST_SCHEMA = "trwm.projection_manifest.v1"


@dataclass(frozen=True)
class ProjectionContract:
    required_fields: tuple[str, ...]
    contract_id: str = "projection.contract"
    schema_version: str = PROJECTION_CONTRACT_SCHEMA

    def __post_init__(self) -> None:
        fields = _normalize_field_names(self.required_fields)
        if not fields:
            raise ValueError("required_fields must not be empty")
        object.__setattr__(self, "required_fields", fields)
        if not self.contract_id:
            raise ValueError("contract_id must be non-empty")
        if self.schema_version != PROJECTION_CONTRACT_SCHEMA:
            raise ValueError(f"unsupported projection contract schema: {self.schema_version}")


@dataclass(frozen=True)
class ProjectionManifest:
    projector_id: str
    projector_version: str
    source_hash: str
    covered_fields: tuple[str, ...]
    field_hashes: Mapping[str, str]
    projection_hash: str = ""
    schema_version: str = PROJECTION_MANIFEST_SCHEMA

    def __post_init__(self) -> None:
        covered = _normalize_field_names(self.covered_fields)
        hashes = {str(key): str(value) for key, value in self.field_hashes.items()}
        object.__setattr__(self, "covered_fields", covered)
        object.__setattr__(self, "field_hashes", hashes)
        if not self.projector_id:
            raise ValueError("projector_id must be non-empty")
        if not self.projector_version:
            raise ValueError("projector_version must be non-empty")
        if self.schema_version != PROJECTION_MANIFEST_SCHEMA:
            raise ValueError(f"unsupported projection manifest schema: {self.schema_version}")
        if set(covered) != set(hashes):
            raise ValueError("field_hashes must match covered_fields")
        expected = _projection_manifest_hash(
            self.schema_version,
            self.projector_id,
            self.projector_version,
            self.source_hash,
            covered,
            hashes,
        )
        if not self.projection_hash:
            object.__setattr__(self, "projection_hash", expected)


@dataclass(frozen=True)
class ProjectionAudit:
    accepted: bool
    missing_fields: tuple[str, ...] = ()
    stale_fields: tuple[str, ...] = ()
    absent_source_fields: tuple[str, ...] = ()
    hash_mismatch: bool = False
    source_hash_mismatch: bool = False

    @property
    def residual(self) -> dict[str, Any]:
        return {
            "kind": "projection_contract_violation",
            "missing_fields": self.missing_fields,
            "stale_fields": self.stale_fields,
            "absent_source_fields": self.absent_source_fields,
            "hash_mismatch": self.hash_mismatch,
            "source_hash_mismatch": self.source_hash_mismatch,
        }


def build_projection_manifest(
    source: Mapping[str, Any],
    covered_fields: Iterable[str],
    *,
    projector_id: str,
    projector_version: str,
) -> ProjectionManifest:
    fields = _normalize_field_names(tuple(covered_fields))
    missing = tuple(field for field in fields if field not in source)
    if missing:
        raise ValueError(f"covered fields not present in source: {missing}")
    field_hashes = {field: field_value_hash(field, source[field]) for field in fields}
    return ProjectionManifest(
        projector_id=projector_id,
        projector_version=projector_version,
        source_hash=source_projection_hash(source),
        covered_fields=fields,
        field_hashes=field_hashes,
    )


def validate_projection_contract(
    contract: ProjectionContract,
    manifest: ProjectionManifest | Mapping[str, Any],
    source: Mapping[str, Any] | None = None,
) -> ProjectionAudit:
    normalized = manifest if isinstance(manifest, ProjectionManifest) else projection_manifest_from_mapping(manifest)
    covered = set(normalized.covered_fields)
    missing = tuple(field for field in contract.required_fields if field not in covered)
    expected_hash = _projection_manifest_hash(
        normalized.schema_version,
        normalized.projector_id,
        normalized.projector_version,
        normalized.source_hash,
        normalized.covered_fields,
        normalized.field_hashes,
    )
    hash_mismatch = bool(normalized.projection_hash and normalized.projection_hash != expected_hash)
    source_hash_mismatch = False
    absent_source_fields: tuple[str, ...] = ()
    stale_fields: tuple[str, ...] = ()
    if source is not None:
        source_hash_mismatch = normalized.source_hash != source_projection_hash(source)
        absent_source_fields = tuple(field for field in contract.required_fields if field not in source)
        stale_fields = tuple(
            field
            for field in contract.required_fields
            if field in source
            and field in normalized.field_hashes
            and normalized.field_hashes[field] != field_value_hash(field, source[field])
        )
    accepted = not missing and not stale_fields and not absent_source_fields and not hash_mismatch and not source_hash_mismatch
    return ProjectionAudit(
        accepted=accepted,
        missing_fields=missing,
        stale_fields=stale_fields,
        absent_source_fields=absent_source_fields,
        hash_mismatch=hash_mismatch,
        source_hash_mismatch=source_hash_mismatch,
    )


def projection_manifest_from_mapping(data: Mapping[str, Any]) -> ProjectionManifest:
    return ProjectionManifest(
        projector_id=str(data["projector_id"]),
        projector_version=str(data["projector_version"]),
        source_hash=str(data["source_hash"]),
        covered_fields=tuple(str(field) for field in data["covered_fields"]),
        field_hashes={str(key): str(value) for key, value in data["field_hashes"].items()},
        projection_hash=str(data.get("projection_hash", "")),
        schema_version=str(data.get("schema_version", PROJECTION_MANIFEST_SCHEMA)),
    )


def source_projection_hash(source: Mapping[str, Any]) -> str:
    return stable_hash({"schema_version": PROJECTION_MANIFEST_SCHEMA, "source": dict(source)})


def field_value_hash(field: str, value: Any) -> str:
    return stable_hash({"field": str(field), "value": value})


def _normalize_field_names(fields: Iterable[str]) -> tuple[str, ...]:
    normalized = tuple(str(field) for field in fields)
    if any(not field for field in normalized):
        raise ValueError("field names must be non-empty")
    deduped = tuple(sorted(set(normalized)))
    if len(deduped) != len(normalized):
        raise ValueError("field names must be unique")
    return deduped


def _projection_manifest_hash(
    schema_version: str,
    projector_id: str,
    projector_version: str,
    source_hash: str,
    covered_fields: tuple[str, ...],
    field_hashes: Mapping[str, str],
) -> str:
    return stable_hash(
        {
            "schema_version": schema_version,
            "projector_id": projector_id,
            "projector_version": projector_version,
            "source_hash": source_hash,
            "covered_fields": covered_fields,
            "field_hashes": dict(field_hashes),
        }
    )

File: src/projection.ts
import { compareCodePoint, stableHash } from "./canonical.js";

export const PROJECTION_CONTRACT_SCHEMA = "trwm.projection_contract.v1";
export const PROJECTION_MANIFEST_SCHEMA = "trwm.projection_manifest.v1";

export interface ProjectionContract {
  requiredFields: string[];
  contractId: string;
  schemaVersion: typeof PROJECTION_CONTRACT_SCHEMA;
}

export interface ProjectionManifest {
  projectorId: string;
  projectorVersion: string;
  sourceHash: string;
  coveredFields: string[];
  fieldHashes: Record<string, string>;
  projectionHash: string;
  schemaVersion: typeof PROJECTION_MANIFEST_SCHEMA;
}

export interface ProjectionAudit {
  accepted: boolean;
  missingFields: string[];
  staleFields: string[];
  absentSourceFields: string[];
  hashMismatch: boolean;
  sourceHashMismatch: boolean;
  residual: {
    kind: "projection_contract_violation";
    missingFields: string[];
    staleFields: string[];
    absentSourceFields: string[];
    hashMismatch: boolean;
    sourceHashMismatch: boolean;
  };
}

export function makeProjectionContract(
  requiredFields: string[],
  options: { contractId?: string } = {},
): ProjectionContract {
  const fields = normalizeFieldNames(requiredFields);
  if (fields.length === 0) {
    throw new RangeError("requiredFields must not be empty");
  }
  const contractId = options.contractId ?? "projection.contract";
  if (!contractId) {
    throw new RangeError("contractId must be non-empty");
  }
  return {
    requiredFields: fields,
    contractId,
    schemaVersion: PROJECTION_CONTRACT_SCHEMA,
  };
}

export async function buildProjectionManifest(
  source: Record<string, unknown>,
  coveredFields: string[],
  options: { projectorId: string; projectorVersion: string },
): Promise<ProjectionManifest> {
  const fields = normalizeFieldNames(coveredFields);
  const missing = fields.filter((field) => !(field in source));
  if (missing.length > 0) {
    throw new RangeError(`covered fields not present in source: ${missing.join(",")}`);
  }
  const fieldHashes: Record<string, string> = {};
  for (const field of fields) {
    fieldHashes[field] = await fieldValueHash(field, source[field]);
  }
  const sourceHash = await sourceProjectionHash(source);
  const projectionHash = await projectionManifestHash({
    schemaVersion: PROJECTION_MANIFEST_SCHEMA,
    projectorId: options.projectorId,
    projectorVersion: options.projectorVersion,
    sourceHash,
    coveredFields: fields,
    fieldHashes,
  });
  return {
    projectorId: options.projectorId,
    projectorVersion: options.projectorVersion,
    sourceHash,
    coveredFields: fields,
    fieldHashes,
    projectionHash,
    schemaVersion: PROJECTION_MANIFEST_SCHEMA,
  };
}

export async function validateProjectionContract(
  contract: ProjectionContract,
  manifest: ProjectionManifest | Record<string, unknown>,
  source?: Record<string, unknown>,
): Promise<ProjectionAudit> {
  const normalized = normalizeProjectionManifest(manifest);
  const covered = new Set(normalized.coveredFields);
  const missingFields = contract.requiredFields.filter((field) => !covered.has(field));
  const expectedHash = await projectionManifestHash(normalized);
  const hashMismatch = Boolean(normalized.projectionHash && normalized.projectionHash !== expectedHash);
  let sourceHashMismatch = false;
  let absentSourceFields: string[] = [];
  const staleFields: string[] = [];
  if (source) {
    sourceHashMismatch = normalized.sourceHash !== await sourceProjectionHash(source);
    absentSourceFields = contract.requiredFields.filter((field) => !(field in source));
    for (const field of contract.requiredFields) {
      if (field in source && field in normalized.fieldHashes) {
        const expectedFieldHash = await fieldValueHash(field, source[field]);
        if (normalized.fieldHashes[field] !== expectedFieldHash) {
          staleFields.push(field);
        }
      }
    }
  }
  const accepted = missingFields.length === 0
    && staleFields.length === 0
    && absentSourceFields.length === 0
    && !hashMismatch
    && !sourceHashMismatch;
  const residual = {
    kind: "projection_contract_violation" as const,
    missingFields,
    staleFields,
    absentSourceFields,
    hashMismatch,
    sourceHashMismatch,
  };
  return {
    accepted,
    missingFields,
    staleFields,
    absentSourceFields,
    hashMismatch,
    sourceHashMismatch,
    residual,
  };
}

export function normalizeProjectionManifest(manifest: ProjectionManifest | Record<string, unknown>): ProjectionManifest {
  const raw = manifest as Record<string, unknown>;
  const schemaVersion = String(raw.schemaVersion ?? raw.schema_version ?? PROJECTION_MANIFEST_SCHEMA);
  if (schemaVersion !== PROJECTION_MANIFEST_SCHEMA) {
    throw new RangeError(`unsupported projection manifest schema: ${schemaVersion}`);
  }
  const projectorId = String(raw.projectorId ?? raw.projector_id ?? "");
  const projectorVersion = String(raw.projectorVersion ?? raw.projector_version ?? "");
  if (!projectorId || !projectorVersion) {
    throw new RangeError("projector id and version must be non-empty");
  }
  const coveredRaw = raw.coveredFields ?? raw.covered_fields;
  if (!Array.isArray(coveredRaw)) {
    throw new RangeError("coveredFields must be an array");
  }
  const fieldHashesRaw = raw.fieldHashes ?? raw.field_hashes;
  if (!fieldHashesRaw || typeof fieldHashesRaw !== "object" || Array.isArray(fieldHashesRaw)) {
    throw new RangeError("fieldHashes must be an object");
  }
  const coveredFields = normalizeFieldNames(coveredRaw.map((field) => String(field)));
  const fieldHashes = Object.fromEntries(
    Object.entries(fieldHashesRaw as Record<string, unknown>)
      .map(([key, value]) => [String(key), String(value)])
      .sort((a, b) => compareCodePoint(a[0], b[0])),
  );
  if (coveredFields.length !== Object.keys(fieldHashes).length || coveredFields.some((field) => !(field in fieldHashes))) {
    throw new RangeError("fieldHashes must match coveredFields");
  }
  return {
    projectorId,
    projectorVersion,
    sourceHash: String(raw.sourceHash ?? raw.source_hash ?? ""),
    coveredFields,
    fieldHashes,
    projectionHash: String(raw.projectionHash ?? raw.projection_hash ?? ""),
    schemaVersion: PROJECTION_MANIFEST_SCHEMA,
  };
}

export async function sourceProjectionHash(source: Record<string, unknown>): Promise<string> {
  return stableHash({ schemaVersion: PROJECTION_MANIFEST_SCHEMA, source });
}

export async function fieldValueHash(field: string, value: unknown): Promise<string> {
  return stableHash({ field: String(field), value });
}

async function projectionManifestHash(manifest: {
  schemaVersion: string;
  projectorId: string;
  projectorVersion: string;
  sourceHash: string;
  coveredFields: string[];
  fieldHashes: Record<string, string>;
}): Promise<string> {
  return stableHash({
    schemaVersion: manifest.schemaVersion,
    projectorId: manifest.projectorId,
    projectorVersion: manifest.projectorVersion,
    sourceHash: manifest.sourceHash,
    coveredFields: manifest.coveredFields,
    fieldHashes: manifest.fieldHashes,
  });
}

function normalizeFieldNames(fields: string[]): string[] {
  const normalized = fields.map((field) => String(field));
  if (normalized.some((field) => !field)) {
    throw new RangeError("field names must be non-empty");
  }
  const deduped = Array.from(new Set(normalized)).sort(compareCodePoint);
  if (deduped.length !== normalized.length) {
    throw new RangeError("field names must be unique");
  }
  return deduped;
}
